src/common/eeprom_loadsave.c

```diff
@@ -0,0 +1,101 @@
+// eeprom_loadsave.c
+
+#include "eeprom_loadsave.h"
+#include <stdio.h>
+#include <string.h>
+#include "config.h"
+#include "st25dv64k.h"
+#include "dbg.h"
+#include "cmsis_os.h"
+#include "ff.h"
+#include "crc32.h"
+
+// public functions for load/save
+
+int eeprom_load_bin_from_usb(const char *fn) {
+    FIL fil;
+    uint8_t buff[128];
+    uint16_t size = 0x800;
+    uint16_t addr = 0x000;
+    UINT bs = sizeof(buff);
+    UINT br;
+    if (f_open(&fil, fn, FA_READ) == FR_OK) {
+        //        eeprom_lock();
+        while (size) {
+            if (size < bs)
+                bs = size;
+            if (f_read(&fil, buff, bs, &br) != FR_OK)
+                break;
+            if (br != bs)
+                break;
+            st25dv64k_user_write_bytes(addr, buff, bs);
+            size -= bs;
+            addr += bs;
+        }
+        f_close(&fil);
+        //        eeprom_unlock();
+        return (size == 0) ? 1 : 0;
+    }
+    return 0;
+}
+
+int eeprom_save_bin_to_usb(const char *fn) {
+    FIL fil;
+    uint8_t buff[128];
+    uint16_t size = 0x800;
+    uint16_t addr = 0x000;
+    UINT bs = sizeof(buff);
+    UINT bw;
+    if (f_open(&fil, fn, FA_WRITE | FA_CREATE_ALWAYS) == FR_OK) {
+        //        eeprom_lock();
+        while (size) {
+            if (size < bs)
+                bs = size;
+            st25dv64k_user_read_bytes(addr, buff, bs);
+            if (f_write(&fil, buff, bs, &bw) != FR_OK)
+                break;
+            if (bw != bs)
+                break;
+            size -= bs;
+            addr += bs;
+        }
+        f_close(&fil);
+        //        eeprom_unlock();
+        return (size == 0) ? 1 : 0;
+    }
+    return 0;
+}
+
+int eeprom_load_xml_from_usb(const char *fn) {
+    return 0;
+}
+
+int eeprom_save_xml_to_usb(const char *fn) {
+    FIL fil;
+    uint8_t id;
+    char text[128];
+    variant8_t var8;
+    UINT bw;
+    uint8_t var_count = eeprom_get_var_count();
+    const char *var_name;
+    if (f_open(&fil, fn, FA_WRITE | FA_CREATE_ALWAYS) == FR_OK) {
+        f_write(&fil, "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n", 55, &bw);
```

Comment:
the attribute standalone has the default value of "no"

Reply:
xml serialization will be improved/finalized in some next pull request, save to xml is now implemented just for debugging and code will be completely reworked.